raft-otel/README.md

```diff
@@ -0,0 +1,112 @@
+
+Distributed consensus is a key component of widely-used software infrastructure such as Zookeeper (used by Kafka), etcd (used by Kubernetes), and the Consul service catalog.
+
+RAFT is a specific protocol that implements distributed consensus. It has been designed specifically to be easier
+to understand than earlier protocols (such as Paxos). However, it remains quite a complex protocol. The [RAFT website](https://raft.github.io) provides 
+a set of resources for understanding how RAFT works. Read the [RAFT paper](https://raft.github.io/raft.pdf), and 
+watch one of the introductory RAFT talks linked from the [RAFT website](https://raft.github.io).
+
+### Background on Distributed Tracing and Open Telemetry
+
+In [other projects](https://github.com/CodeYourFuture/immersive-go-course/tree/main/kafka-cron) we have added metrics to our programs and collected and 
+aggregated those metrics using the Prometheus monitoring tool. Metrics are a widely-used methodology for understanding the behaviour of our systems at 
+a statistical level: what percentage of requests are being completed successfully, what is the 90th percentile latency, what is our current cache hit 
+rate or queue length. These kinds of queries are very useful for telling us whether our systems seem to be healthy overall or not, and, in some cases, 
+may provide useful insights into problems or inefficiencies.
+
+However, one thing that metrics are not normally very good for is understanding how user experience for a system may vary between different types of 
+requests, why particular requests are outliers in terms of latency, and how a single user request flows through backend services - many complex web services
+may involve dozens of backend services or datastores. It may be possible to answer some of these questions using logs analysis. 
+However, there is a better solution, designed just for this problem: distributed tracing. 
+
+Distributed tracing has two key concepts: traces and spans. A trace represents a whole request or transaction. Traces are uniquely identified by trace IDs.
+Traces are made up of a set of spans, each tagged with the trace ID of the trace it belongs to. Each span is a unit of work: a remote procedure call or 
+web request to a specific service, a method execution, or perhaps the time that a message spends in a queue. Spans can have child spans. 
+There are specific tools that are designed to collect and store distributed traces, and to perform useful queries against them. 
+One of the key aspects of distributed tracing is that when services call other services the trace ID is propagated to those calls, so that the 
```

Comment:
I think it's probably worth explicitly mentioning that each service will typically independently post spans to a central collector. If you've never encountered the concept before, I can see how not having that spelled out may leave someone confused, or thinking spans accumulate system-to-system through calls...

Reply:
done, fair point